pkg/webui/console/containers/device-onboarding-form/device-type-form-section/device-type-manual-form-section/validation-schema.js

```diff
@@ -12,14 +12,148 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import { isUndefined } from 'lodash'
+
+import sharedMessages from '@ttn-lw/lib/shared-messages'
 import Yup from '@ttn-lw/lib/yup'
 
-// Validation schema of the device type manual form section.
+import { ACTIVATION_MODES } from '@console/lib/device-utils'
+
+import { DEVICE_CLASS_MAP } from '../../utils'
+
+// Validation schemas of the device type manual form section.
 // Please observe the following rules to keep the validation schemas maintainable:
 // 1. DO NOT USE ANY TYPE CONVERSIONS HERE. Use decocer/encoder on field level instead.
 //    Consider all values as backend values. Exceptions may apply in consideration.
 // 2. Comment each individual validation prop and use whitespace to structure visually.
 // 3. Do not use ternary assignments but use plain if statements to ensure clarity.
-const validationSchema = Yup.object({})
+
+const factoryPresetFreqRequiredTest = frequencies =>
+  frequencies.every(freq => typeof freq !== 'undefined' && freq !== '')
+
+const factoryPresetFreqNumericTest = frequencies =>
+  frequencies.every(freq => {
+    if (typeof freq !== 'undefined') {
+      return !isNaN(parseInt(freq))
+    }
+
+    return true
+  })
+
+const advancedSettingsSchema = Yup.object({
+  supports_class_b: Yup.boolean(),
+  supports_class_c: Yup.boolean(),
+  supports_join: Yup.boolean(),
+  multicast: Yup.boolean(),
+  _device_class: Yup.string().when(['_activation_mode'], (mode, schema) => {
+    // The additional device classes (B and/or C) have to be set when using multicast.
+    if (mode === ACTIVATION_MODES.MULTICAST) {
+      return schema.required(sharedMessages.validateRequired)
+    }
+
+    return schema.oneOf(Object.values(DEVICE_CLASS_MAP))
+  }),
+  _default_ns_settings: Yup.bool(),
+  _activation_mode: Yup.mixed().when(['$mayEditKeys'], (mayEditKeys, schema) => {
+    // If the user may not set any root keys, the only activation option is OTAA.
+    if (!mayEditKeys) {
+      return schema.oneOf([ACTIVATION_MODES.OTAA]).required(sharedMessages.validateRequired)
+    }
+
+    return schema.oneOf(Object.values(ACTIVATION_MODES)).required(sharedMessages.validateRequired)
+  }),
+  _skip_js_registration: Yup.bool(),
```

Comment:
Note how validation schemas become a lot cleaner now that they can focus on validation only and not on any transformation logic.